bulkio: Support running BACKUP under transaction.

miretskiy opened the pull request:
Informs #47539

Allow execution of BACKUP statement under explicit transaction,
provided the "DETACHED" option is specified:

```
BACKUP TO <location_uri> ... WITH DETACHED
```

When backup runs in `detached` mode, we do not wait for the job
completion.  Instead, we return the job id to the caller, and
the job executes asynchronously.

Release notes (enterprise change): BACKUP can now run in "detached" mode.
That is, we do not wait for the BACKUP job to finish.  Instead, the
job ID is returned immediately, and the job itself runs in background.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50775)
<!-- Reviewable:end -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50775)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/backupccl/backup_test.go, line 3473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4mWp4Ay1DXtA9RwOJ:-MB4mWp4Ay1DXtA9RwOK:bv3us1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/ccl/backupccl/backup_test.go#L3473)):*
> ```Go
> }
> 
> func waitForSuccessfulJob(t *testing.T, tc *testcluster.TestCluster, id int64) {
> ```

By the way, we have something like this in `jobsutil.WaitForJob` that we use in some of the backup tests as well as import and create stats test.

Could we make that method nudge the registry and re-use it?
___
*[pkg/ccl/backupccl/backup_test.go, line 3492 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4ocPV6h8waUMw82bE:-MB4ocPV6h8waUMw82bF:b-1rc2p3) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/ccl/backupccl/backup_test.go#L3492)):*
> ```Go
> 	db := sqlDB.DB.(*gosql.DB)
> 
> 	// running backup under transaction requires DETACHED
> ```

nit: comment sentence formatting 
___
*[pkg/ccl/backupccl/backup_test.go, line 3507 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4o4hh2Tjb52w-zkws:-MB4o4hiA3VANmyCOsY_:byh5qvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/ccl/backupccl/backup_test.go#L3507)):*
> ```Go
> 	// Backup again, under explicit transaction.
> 	tx, err = db.Begin()
> 	require.NoError(t, err)
> ```

Thanks for cleaning these up!
___
*[pkg/ccl/backupccl/backup_test.go, line 3511 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4pVwr123PLBD2vEUc:-MB4pVwr123PLBD2vEUd:b-ucdzfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/ccl/backupccl/backup_test.go#L3511)):*
> ```Go
> 	require.NoError(t, err)
> 	require.NoError(t, tx.Commit())
> 	waitForSuccessfulJob(t, tc, jobID)
> ```

Can we also add a test for rolling back the transaction instead of committing?
___
*[pkg/jobs/registry.go, line 383 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4nsE38BSZlPr2w5aJ:-MB4nsE4AMU22LcpPBzf:b-vq14vu) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/jobs/registry.go#L383)):*
> ```Go
> 
> // CreateAdoptableJobWithTxn creates a job which will be adopted for execution at a later time
> // by some node in a cluster.
> ```

nit: s/in a cluster/in the cluster/
___
*[pkg/jobs/registry.go, line 388 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4n_7T3T9I5vLd2zTF:-MB4n_7T3T9I5vLd2zTG:b4obdd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/jobs/registry.go#L388)):*
> ```Go
> ) (*Job, error) {
> 	j := r.NewJob(record)
> 	if err := j.WithTxn(txn).insert(
> ```

I think it's worth adding a comment explicitly mentioning that we purposely create a lease with an invalid node ID so that the registry will adopt it later.
___
*[pkg/jobs/registry.go, line 1033 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4ljcm44BHgjQDO8Cy:-MB4ljcm44BHgjQDO8Cz:b-c4je6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/jobs/registry.go#L1033)):*
> ```Go
> 	}
> 	nodeStatusMap := map[roachpb.NodeID]*nodeStatus{
> 		// 0 is not a valid node ID, but we treat it as an always-dead node so that
> ```

should we update this comment since it's not clear what `0` is referring to anymore? Something like "We treat invalidNodeID as an ..."


<!-- Sent from Reviewable.io -->

pbardea commented:
Also wanted to flag that it looks like there are a couple of proto generated files included here - probably due to the Go version upgrade.

Also, I don't think "improvement" is a release category. I think the closest category would be "enterprise change".

Commit:
bulkio: Support running BACKUP under transaction.

Allow execution of BACKUP statement under explicit transaction,
provided the "DETACHED" option is specified:

```
BACKUP TO <location_uri> ... WITH DETACHED
```

When backup runs in `detached` mode, we do not wait for the job
completion.  Instead, we return the job id to the caller, and
the job executes asynchronously.

Release notes (enterprise change): BACKUP can now run in "detached" mode.
That is, we do not wait for the BACKUP job to finish.  Instead, the
job ID is returned immediately, and the job itself runs in background.

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50775)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/backupccl/backup_test.go, line 3473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4mWp4Ay1DXtA9RwOJ:-MBAygHNCjLoNmpbvOfZ:b5d8y6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/ccl/backupccl/backup_test.go#L3473)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

By the way, we have something like this in `jobsutil.WaitForJob` that we use in some of the backup tests as well as import and create stats test.

Could we make that method nudge the registry and re-use it?
</blockquote></details>

I saw that... But, tbh, I wasn't comfortable adding nudge to that utility; and if I'm going to write a wrapper right here (to nudge and call jobsutil.WaitForJob), then I thought that it might actually be more readable if this happened inline, right here.  If you feel strongly, I can switch.
___
*[pkg/ccl/backupccl/backup_test.go, line 3507 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4o4hh2Tjb52w-zkws:-MBAzmgP6GyD7jHcgsun:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/ccl/backupccl/backup_test.go#L3507)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Thanks for cleaning these up!
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 388 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4n_7T3T9I5vLd2zTF:-MBB1arB31IHyn-VN3ty:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/jobs/registry.go#L388)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

I think it's worth adding a comment explicitly mentioning that we purposely create a lease with an invalid node ID so that the registry will adopt it later.
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 1033 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4ljcm44BHgjQDO8Cy:-MBB1piN8C0F_BYM2h6d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/jobs/registry.go#L1033)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

should we update this comment since it's not clear what `0` is referring to anymore? Something like "We treat invalidNodeID as an ..."
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50775)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/ccl/backupccl/backup_test.go, line 3511 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50775#-MB4pVwr123PLBD2vEUc:-MBBVxcIB5q3TaWrZTxu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/368aa390e0d71bab6e1123501871c33beac622db/pkg/ccl/backupccl/backup_test.go#L3511)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Can we also add a test for rolling back the transaction instead of committing?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy commented:
@pbardea @dt ready for another look;

pbardea commented on pkg/jobs/registry.go:
```diff
@@ -377,6 +377,25 @@ func (r *Registry) CreateJobWithTxn(ctx context.Context, record Record, txn *kv.
 
+const invalidNodeID = 0
+
+// CreateAdoptableJobWithTxn creates a job which will be adopted for execution at a later time
```
super minor nit: it looks like this comment and the one below are wrapped to 100 char, rather than the usual 80

miretskiy replied:
Done

dt commented on pkg/ccl/utilccl/jobutils.go:
```diff
@@ -0,0 +1,51 @@
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+)
+
+// JobExecutionResultHeader is the header for various job commands
```
nit: I'm not sure that there is a common _result_ header across all job types, or even if we wanted one, it'd say it is separate from the _queued_ job work, so i might skip this?

miretskiy replied:
Sorry, @dt... Just started bors.  I broke out bigger PR where I had detached for all of restore, backup and import.  All of them used the same headers.  That's why I moved it to the utility.

miretskiy replied:
Let me try to renamed it.  BulkJobExecutionResultHeader

dt replied:
It isn't even common to the bulk jobs though -- import and restore have different heads. I'd skip it entirely.

miretskiy commented:
bors r+

miretskiy commented:
bors r-

miretskiy commented:
bors r+

miretskiy commented:
bors r+

miretskiy commented:
bors r+

dt commented:
closes #47539?

miretskiy commented:
Closes #47539

miretskiy commented:
yup

On Thu, Jul 2, 2020 at 8:06 PM David Taylor <notifications@github.com>
wrote:

> closes #47539 <https://github.com/cockroachdb/cockroach/issues/47539>?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/50775#issuecomment-653267061>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVB6QPOVATAPFU6YFMLRZUOJHANCNFSM4OLQTRUQ>
> .
>
